### PR TITLE
[Feature][JsonSplit] add task delete/update/switch method

### PR DESCRIPTION
**Pull request by JinyLeeChina**
add task delete/update/switch method #4417

### CODE REVIEW

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java [hunk 1]
```diff
@@ -100,14 +111,8 @@
...
         TaskNode taskNode = JSONUtils.parseObject(taskDefinitionJson, TaskNode.class);
-        if (taskNode == null) {
-            logger.error("taskDefinitionJson is not valid json");
-            putMsg(result, Status.DATA_IS_NOT_VALID, taskDefinitionJson);
-            return result;
-        }
-        if (!CheckUtils.checkTaskNodeParameters(taskNode.getParams(), taskNode.getName())) {
-            logger.error("task node {} parameter invalid", taskNode.getName());
-            putMsg(result, Status.PROCESS_NODE_S_PARAMETER_INVALID, taskNode.getName());
+        checkTaskNode(result, taskNode, taskDefinitionJson);
```
Comment (lenboo) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java:114]: missing 'result = ？'

**Review by JinyLeeChina — COMMENTED** [review]
result is a map, which is passed into the method through the parameter

**Comment by JinyLeeChina** [comment]
> result is a map, which is passed into the method through the parameter



> result is a map, which is passed into the method through the parameter

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java [hunk 2]
```diff
@@ -204,5 +211,188 @@ private String getResourceIds(TaskDefinition taskDefinition) {
...
+        taskDefinition.setTaskPriority(taskDefinitionLog.getTaskPriority());
+        taskDefinition.setWorkerGroup(taskDefinitionLog.getWorkerGroup());
+        taskDefinition.setFailRetryTimes(taskDefinitionLog.getFailRetryTimes());
+        taskDefinition.setFailRetryInterval(taskDefinitionLog.getFailRetryInterval());
+        taskDefinition.setTimeoutFlag(taskDefinitionLog.getTimeoutFlag());
+        taskDefinition.setTaskTimeoutStrategy(taskDefinitionLog.getTaskTimeoutStrategy());
+        taskDefinition.setTimeout(taskDefinitionLog.getTimeout());
+        taskDefinition.setUpdateTime(new Date());
+        taskDefinition.setResourceIds(taskDefinitionLog.getResourceIds());
+        taskDefinitionMapper.updateById(taskDefinition);
```
Comment (simon824) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java:392]: missing setId or updateByCode here?

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java [hunk 3]
```diff
@@ -204,5 +211,188 @@ private String getResourceIds(TaskDefinition taskDefinition) {
...
+        taskDefinition.setTaskPriority(taskNode.getTaskInstancePriority());
+        taskDefinition.setWorkerGroup(taskNode.getWorkerGroup());
+        taskDefinition.setFailRetryTimes(taskNode.getMaxRetryTimes());
+        taskDefinition.setFailRetryInterval(taskNode.getRetryInterval());
+        taskDefinition.setTimeoutFlag(taskNode.getTaskTimeoutParameter().getEnable() ? TimeoutFlag.OPEN : TimeoutFlag.CLOSE);
+        taskDefinition.setTaskTimeoutStrategy(taskNode.getTaskTimeoutParameter().getStrategy());
+        taskDefinition.setTimeout(taskNode.getTaskTimeoutParameter().getInterval());
+        taskDefinition.setUpdateTime(now);
+        taskDefinition.setResourceIds(getResourceIds(taskDefinition));
+        taskDefinitionMapper.updateById(taskDefinition);
```
Comment (simon824) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java:304]: missing setId or updateByCode here?

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessTaskRelationService.java [hunk 4]
```diff
@@ -48,5 +48,77 @@
...
+     * delete process task relation
+     *
+     * @param loginUser login user
+     * @param projectName project name
+     * @param processDefinitionCode process definition code
+     */
+    Map<String, Object> deleteTaskDefinitionByProcess(User loginUser,
+                                                      String projectName,
+                                                      Long processDefinitionCode);
+
```
Comment (simon824) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessTaskRelationService.java:73]: The function name deleteTaskDefinitionByProcessCode maybe more easy to understand

Comment (JinyLeeChina) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessTaskRelationService.java:73]: OK

**Review by lenboo — APPROVED** [review]
+1

**Pull request merged**